### ISSUE TITLE
[Dialog] Close contextual elements with the `Esc` key in visual order

### ISSUE THREAD

**Issue by geospatialem** [Esri/calcite-design-system#6456]
### Description
Provide similar behavior to keyboard functionality to mouse users where if multiple modal dialogs are opened, the Esc key provides similar functionality to when selecting the components close button.

cc @Anastasiia-Boleiko

### Acceptance Criteria
If multiple modal dialog windows are open and the Esc key is pressed, the most recently opened (highest order, or visually prominent) should be closed first. The order should follow until the last dialog is closed.

### Relevant Info
Came up in #6292 as a nice to have, but was out of scope for the bug fix.

### Which Component
modal

### Example Use Case
https://codepen.io/anastasiia-zrh/pen/MWBvZdB

### Esri team
ArcGIS Urban

**Comment by cont-kolomeets** [comment]
Another case related to this one is to close an open popover first on the 1st Esc press, then close the modal on the 2nd Esc press.
Right now hitting Esc closes both at once.

### Example Use Case

https://codepen.io/cont-kolomeets/pen/RwqzgqP?editors=100

![image](https://github.com/Esri/calcite-design-system/assets/5318527/58347e06-8224-4080-9456-4281239b68c0)

**Comment by geospatialem** [comment]
Spike to determine if the new `dialog` needs to be updated to handle `modal` with the visual order specified above has the current functionality, or if needed, should be supported with the deprecation of `modal` in the upcoming `2.11.0` release (7/30).

**Comment by geospatialem** [comment]
Adding some notes @Elijbet compiled for testing out the functionality in the [attached PR](https://github.com/Esri/calcite-design-system/pull/9231#issuecomment-2397808476) ✨💪🏻✨

**Comment by DitwanP** [comment]
🍡 Verified on `2.14.0-next.18` 

https://codepen.io/Ditwan-Price/pen/MWNQdvw?editors=1011
